### PR TITLE
[LLVM][AsmParser] Add support for C style comments

### PR DESCRIPTION
**Pull request by jurahul**
Add support for C style comments in LLVM assembly.

### CODE REVIEW

File: llvm/lib/AsmParser/LLLexer.cpp [hunk 1]
```diff
@@ -262,6 +274,38 @@ void LLLexer::SkipLineComment() {
...
+/// is that we allow nesting.
+bool LLLexer::SkipCComment() {
+  getNextChar(); // skip the star.
+  unsigned CommentDepth = 1;
+
+  while (true) {
+    int CurChar = getNextChar();
+    switch (CurChar) {
+    case EOF:
+      LexError("Unterminated comment!");
```
Comment (arsenm) [inline llvm/lib/AsmParser/LLLexer.cpp:287]: Messages should start with lowercase and be less excited

Comment (jurahul) [reply llvm/lib/AsmParser/LLLexer.cpp]: Done.

**Comment by jurahul** [comment]
@arsenm any further comments? Are we at this point waiting for some consensus to be reached around the RFC?

**Comment by arsenm** [comment]
This is pretty independent of the RFC

**Comment by jurahul** [comment]
It is. Lets wait for @nikic to review as well.

File: llvm/lib/AsmParser/LLLexer.cpp [hunk 2]
```diff
@@ -262,6 +274,38 @@ void LLLexer::SkipLineComment() {
   }
 }
 
+/// SkipCComment - This skips C-style /**/ comments.  The only difference from C
+/// is that we allow nesting.
```
Comment (jrtc27) [inline llvm/lib/AsmParser/LLLexer.cpp:278]: Is there a good reason to support this / are there other languages out there that support nesting C-style comments? I don't think it's a good idea to deviate from C here unless there's a very good reason to.

Comment (topperc) [reply llvm/lib/AsmParser/LLLexer.cpp:278]: This concept appears to have been copied from TGLexer.cpp

Comment (jrtc27) [reply llvm/lib/AsmParser/LLLexer.cpp:278]: Interesting. Dates back to af282ae26e784d42b15ae3aff887f4f3769eddec, July 2003.

**Comment by jurahul** [comment]
Yes, it is adopted from TGLexer. There is no good reason to support nested
comments though.

File: llvm/lib/AsmParser/LLLexer.cpp (hunk 2 above)
Comment (jurahul) [reply llvm/lib/AsmParser/LLLexer.cpp]: Dropped nested comment support. I suspect at that point, TableGen did not support preprocessor macros, so made sense to support nested C style comments to enable commenting out code (?)

**Review by nikic — COMMENTED** [review]
This needs to be documented in LangRef.

**Comment by jurahul** [comment]
> This needs to be documented in LangRef.

Done.

File: llvm/test/Assembler/invalid-c-style-comment1.ll [hunk 3]
```diff
@@ -0,0 +1,10 @@
+; RUN: not llvm-as --disable-output %s 2>&1 | FileCheck %s -DFILE=%s
+
+@B = external global i32
+
+*/
+
+/*   /* Nested comments not supported */
+
+; CHECK: [[FILE]]:[[@LINE+1]]:1: error: redefinition of global '@B'
```
Comment (nikic) [inline llvm/test/Assembler/invalid-c-style-comment1.ll:9]: This is not the error I would have expected for this test case. Why isn't there an error at the first `*/`?

Comment (jurahul) [reply llvm/test/Assembler/invalid-c-style-comment1.ll:9]: WDYM? Even in C: "/* hello /* this is comment */ " is a full comment. The second /* is ignored, and the comment ends at the first */. Clang does give a warnings:

```
comment.c:2:10: warning: '/*' within block comment [-Wcomment]
/* hello /* this is comment */
```

but accepts this code. Are you suggesting that we be strict and disallow /* within the /* */ comment?

Comment (jurahul) [reply llvm/test/Assembler/invalid-c-style-comment1.ll:9]: Looking at Clang Lexer, it has:

```
      // If this is a /* inside of the comment, emit a warning.  Don't do this
      // if this is a /*/, which will end the comment.  This misses cases with
      // embedded escaped newlines, but oh well.```
```

So if we add the `/*` within `/*...*/` comment check, we just need to handle this case of allowing `/*/` to terminate the comment.

File: llvm/test/Assembler/invalid-c-style-comment1.ll [hunk 4]
```diff
@@ -0,0 +1,10 @@
+; RUN: not llvm-as --disable-output %s 2>&1 | FileCheck %s -DFILE=%s
+
+@B = external global i32
+
+*/
```
Comment (nikic) [inline llvm/test/Assembler/invalid-c-style-comment1.ll:5]: I'm referring to the `*/` here, not to the comment below. I'd expect a closing `*/` without opening `/*` to generate an error.

Comment (jurahul) [reply llvm/test/Assembler/invalid-c-style-comment1.ll:5]: You're right, I think you found a bug in the code :) I'll remove tis stray */ here and add another unit test for this.

File: llvm/lib/AsmParser/LLLexer.cpp [hunk 5]
```diff
@@ -251,6 +258,12 @@ lltok::Kind LLLexer::LexToken() {
     case ',': return lltok::comma;
     case '*': return lltok::star;
     case '|': return lltok::bar;
+    case '/':
+      if (peekNextChar() != '*')
```
Comment (jurahul) [inline llvm/lib/AsmParser/LLLexer.cpp:262]: This is the fix. The old code was:

```
      if (peekNextChar() == '*' && SkipCComment())
        return lltok::Error;
      continue;
```

Given the input:

```
@B = external global i32

*/
```

The `@B = external global i32*` is a valid sequence of tokens (The * becomes a part of the i32 * type). The `/` following the * is silently eaten by the old code due to the `continue`. The fix now correctly executes `continue` only when it has successfully parsed a C style comment.

File: llvm/test/Assembler/invalid-c-style-comment2.ll [hunk 6]
```diff
@@ -0,0 +1,7 @@
+; RUN: not llvm-as --disable-output %s 2>&1 | FileCheck %s -DFILE=%s
+
+@B = external global i32
+
+; CHECK: [[FILE]]:[[@LINE+1]]:2: error: expected top-level entity
+*/
```
Comment (jurahul) [inline llvm/test/Assembler/invalid-c-style-comment2.ll:6]: This is the new stand-alone unit test for the bug you noticed.

File: llvm/lib/AsmParser/LLLexer.cpp (hunk 5 above)
Comment (jurahul) [reply llvm/lib/AsmParser/LLLexer.cpp]: Right, Thanks. I have reverted that part of the change and the code is now using getNextChar() directly. That entailed handling EOF following a * within a comment, so added another unit test for that.

File: llvm/lib/AsmParser/LLLexer.cpp [hunk 7]
```diff
@@ -262,6 +267,28 @@ void LLLexer::SkipLineComment() {
   }
 }
 
+/// SkipCComment - This skips C-style /**/ comments. Returns true if there
```
Comment (nikic) [inline llvm/lib/AsmParser/LLLexer.cpp:270]:
```suggestion
/// This skips C-style /**/ comments. Returns true if there
```
Don't repeat function name in doc comments. (This is an old style that shouldn't be replicated.)

**Review by nikic — APPROVED** [review]
LGTM, but please wait for a second approval.

**Comment by jurahul** [comment]
> LGTM, but please wait for a second approval.

Thanks, will do.

**Comment by jurahul** [comment]
@arsenm @jrtc27 can one of you give an additional approval?

**Review by arsenm — APPROVED** [review]

**Comment by jurahul** [comment]
Thanks @arsenm

**Pull request merged**